### PR TITLE
CacheDictionary datarace on exception_ptr 

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix datarace on exception_ptr in CacheDictionary after #8303.

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Test will be added.

File: dbms/src/Common/ProfilingScopedRWLock.h [hunk 1]
```diff
@@ -17,6 +17,17 @@ class ProfilingScopedWriteRWLock
         ProfileEvents::increment(event, watch.elapsed());
     }
 
+    /// Unsafe!
```
Comment (akuzm) [inline dbms/src/Common/ProfilingScopedRWLock.h]: The comment is not helpful, we should explain more or remove it. Looks like this is just a generic lock/unlock method that is not particularly unsafe.

Comment (akuzm) [reply dbms/src/Common/ProfilingScopedRWLock.h]: Also we should increment the metric when we re-lock.

File: dbms/src/Dictionaries/CacheDictionary.cpp [hunk 2]
```diff
@@ -832,9 +835,14 @@ void CacheDictionary::update(BunchUpdateUnit & bunch_update_unit) const
...
+            /// To perform parallel loading.
+            BlockInputStreamPtr stream = nullptr;
+            {
+                write_lock.unlock();
+                stream = source_ptr->loadIds(bunch_update_unit.getRequestedIds());
+                write_lock.lock();
+            }
 
             stream->readPrefix();
             while (const auto block = stream->read())
```
Comment (akuzm) [inline dbms/src/Dictionaries/CacheDictionary.cpp]: read() might block for some storages, so maybe we should release the lock here as well.

File: dbms/src/Dictionaries/CacheDictionary.cpp [hunk 3]
```diff
@@ -832,9 +835,14 @@ void CacheDictionary::update(BunchUpdateUnit & bunch_update_unit) const
...
 
             Stopwatch watch;
-            auto stream = source_ptr->loadIds(bunch_update_unit.getRequestedIds());
 
-            const ProfilingScopedWriteRWLock write_lock{rw_lock, ProfileEvents::DictCacheLockWriteNs};
+            /// To perform parallel loading.
+            BlockInputStreamPtr stream = nullptr;
+            {
+                write_lock.unlock();
+                stream = source_ptr->loadIds(bunch_update_unit.getRequestedIds());
```
Comment (akuzm) [inline dbms/src/Dictionaries/CacheDictionary.cpp]: We can still end up with unlocked lock in catch(), if loadIds throws.

**Comment by nikitamikhaylov** [comment]
No backport needed, because in previous versions we don't have "async update" feature.

**Comment by alexey-milovidov** [comment]
One build has failed due to internal CI issue.

> Functional stateless tests (debug) 

- no space left on device.

**Comment by alexey-milovidov** [comment]
> Integration tests (asan)

- known issue.

File: dbms/src/Common/ProfilingScopedRWLock.h [hunk 4]
```diff
@@ -7,21 +7,51 @@
 
 namespace DB
 {
+
+class ProfilingScoperWriteUnlocker;
```
Comment (alexey-milovidov) [inline dbms/src/Common/ProfilingScopedRWLock.h]: Scoper -> Scoped.

**Review by alexey-milovidov — APPROVED** [review]

File: dbms/src/Common/ProfilingScopedRWLock.h (hunk 4 above)
Comment (alexey-milovidov) [reply dbms/src/Common/ProfilingScopedRWLock.h]: (merge after fix of this typo)

**Comment by KochetovNicolai** [comment]
There is no such code in 20.1

**Pull request merged**